### PR TITLE
Update build process to default to static value builds, use verbose details for dev builds when available

### PR DESCRIPTION
**Pull request by deoren**
This is intended for use with dev builds of the docs to quickly identify what branch the docs were bult from. This also helps with epub builds as some readers do not always display the "release" details prominently once the file has been opened.

Note: Looks like my editor applied some whitespace trimming when I saved the index file, so the commit isn't entirely "clean".

@rgerhards What do you think? The results should be visible here soon (builds are performed every 15 minutes, takes just under 10 minutes to complete):

http://rsyslog.whyaskwhy.org/display_branch_build_info/

### CODE REVIEW

**Comment by deoren** [comment]
Huh. Looks like I still have a kink to work out.

**Comment by deoren** [comment]
Looks like everything works properly if building the docs from an actual branch, but if checking out in a detached head state, then 'HEAD' becomes the branch name. Not intended, so I'll have to do further testing on this in re to the cron build script. In short, it may be a while before this is ready.

**Comment by rgerhards** [comment]
@deoren 

* we need to think if we can provide a way to set the version number for stable builds in a fixed way -- right now I think I have to modify the web server script after each release
* do you think it would make sense to have a "doc-experimental" or whatever branch that we would frequently render on the rsyslog site? So we would see how it looks there as well. But it's a bit clumsy...

**Comment by deoren** [comment]
> @rgerhards: 
>
> * we need to think if we can provide a way to set the version number for stable builds in a fixed way -- right now I think I have to modify the web server script after each release

We should have that functionality now, either manually via:

`sphinx-build -D version="8.32" -D release="8.32.0" -b html source build`

or via the updated (#486) `release_build.sh` script in the master branch. I just checked out a fresh copy of the repo and used that script to build the docs. From what I can tell the results are as expected.

`deoren@testbox:/tmp/rsyslog-doc$ ls -l *.gz`
```
-rw-rw-r-- 1 deoren deoren 5358668 Jan 12 08:59 rsyslog-doc-8.32.tar.gz
```

`deoren@testbox:/tmp/rsyslog-doc$ grep '8.3' build/index.html`
```
    <title>Welcome to Rsyslog &#8212; rsyslog 8.32.0 documentation</title>
        VERSION:     '8.32.0',
        <li class="nav-item nav-item-0"><a href="#">rsyslog 8.32.0 documentation</a> &#187;</li>
        <li class="nav-item nav-item-0"><a href="#">rsyslog 8.32.0 documentation</a> &#187;</li>
```

> @rgerhards: 
> * do you think it would make sense to have a "doc-experimental" or whatever branch that we would frequently render on the rsyslog site? So we would see how it looks there as well. But it's a bit clumsy...

Provided that automated builds from that branch are enabled, it would make it easier/safer for contributors to see how the docs would render within the confines of WordPress integration. The way it is now you have to ask your team to upload/decompress tarballs on an on-demand basis and that manual effort is bound to lag behind changes to the branch. If you did go this route it could be useful to help test new themes and major changes.

However, if we're just wanting to use an established/agreed-upon theme, then I'm fine using my VPS for now. My approach is to build _all_ branches, so that I can easily work on multiple prototypes in parallel. I'm using my setup as a means of working out the kinks before approaching your group about setting up something official to replicate the results for other team members.

In short, I think having a dev/prototype branch has value (PR #476 is a good example) and it is something that we could work into a CONTRIB doc at some point, aimed primarily at regular contributors who are making large changes. 

On a very loosely related note, how long are branches kept in the repo if they've already been merged in? It looks like the `imfile-timeout` branch was merged in a while back, but it's in the repo.

**Comment by deoren** [comment]
@rgerhards On a related note, I plan to setup a Makefile to build the Sphinx documentation. The more I read the Sphinx user docs and poke around other Sphinx-based documentation, I see that the Sphinx quick start process results in the generation of a starter Makefile for further customization. We could use something like that to produce the desired effect.

Examples:

- `make dist` to create release docs, tarball
- `make allbranches` to build all branches
- `make dev` to build master branch

and so forth. Having something like this could replace `release_build.sh` or "drive" it with appropriate options. I'm particularly interested in using this approach to replace (or greatly simplify) the cron script I'm referencing with this PR.

Example Makefiles:

- https://github.com/pallets/werkzeug/blob/master/docs/Makefile
- https://github.com/sphinx-doc/sphinx/blob/master/Makefile

**Comment by deoren** [comment]
> @rgerhards: 
>
> * we need to think if we can provide a way to set the version number for stable builds in a fixed way --

I responded to that in https://github.com/rsyslog/rsyslog-doc/pull/493#issuecomment-357264729, but I initially missed this part:

> right now I think I have to modify the web server script after each release

It sounds like you're using a separate script that maybe I've not seen, perhaps from a private repo?

The `release_build.sh` script doesn't automatically pull the info from Git (though we can add that functionality), but it does take the major.minor version info from command-line and use that to set the version/release values used by the release build of the docs. I think you're much more skilled than I, but if you want to share the script that you're referencing I'll be happy to take a look at it.

I made references to this before, but I'd recommend moving to automated doc builds and automated site content updates. Again, I suspect you're more skilled than I, but I would be happy to share some thoughts on this if you wish.

**Comment by deoren** [comment]
@rgerhards Please disregard all of my previous comments on this PR as much has changed since then.

You of course saw #504 and likely also saw that I submitted PR #505 to work around that.

I had been thinking that changing the default build from a "dev" build to a "stable" build was the direction I needed to go, but #504 decided it for me: breaking downstream maintainer processes is a blocker for any enhancements to the build process.

I went back and reworked some of the improvements I have made on this branch (and have since rebased off of `master`) to default to building a stable build if building from a release tarball generated by the `release_build.sh` script and a "dev" build otherwise.

I'm going to spend a little more time testing the changes before I seek to merge them, but when you have a moment, please do the following:

## Prep

1. Create a fresh clone from https://github.com/deoren/rsyslog-doc
1. Switch to the poorly named `display_branch_build_info` branch
    -  *I would rename the branch, but I'm new enough to Git/GitHub that I'm not sure whether renaming the branch will result in this PR automatically closing*

## Test release build process

1. Run `release_build.sh`
1. Examine the HTML output
1. Copy the tarball to a temp location and decompress it
1. cd to that location (e.g., the 'source' directory can be found at `./source`)
1. Examine the `source/conf.py` file within the tarball
    - *The `conf.py` file is now modified by `release_build.sh` to replace placeholders with the latest stable version*
1. Run `sphinx-build -b html source build`
1. Check the results
    - *The version should be set to '8.32' and the release string set to '8.32.0' as those are associated with the latest stable tag*

## Test dev build process

1. Change back to your clone of my fork
1. Run `sphinx-build -b html source build` directly
1. Examine the HTML output
    1. You should see the verbose release string across the top-left
    1. You should also see a build string at the bottom of the index page indicating the build date, the branch and the commit hash. This is shown for dev builds ONLY and is primarily useful for reviewing alternate formats (e.g., epub on a tablet or phone) that might not prominently display the release string in the same way that the HTML format is displayed.

Provided that I didn't goof up anything between the time I write this and the time you review the branch, I think you'll find that the build process is "usable" now. :P 

Once this has been merged and everything works as expected for a time, I'll turn to setting up a Makefile, but I'd like to make that a separate PR as I feel this one is already complex enough.

Once we have a chance to work out any remaining kinks, I will ask mbiebl to test on his end to make sure what I'm proposing won't cause him any issues.

**Comment by rgerhards** [comment]
@deoren Unfortunately I will not be able to do a fresh test today, but let me at least give my personal opinion (maintainer hat off ;-))

* The web server script I used for the rsyslog site is very simplistic --  a git checkout followed by sphinx-build. Maybe a short doc on what to use instead might be useful not only for me.
* I think we should differentiate between two cases
  - stable builds 
      - need a plain version number, eg. 8.32.0
      - always come from v?-stable git branch or tarball
  - dev (master) builds
    - IMHO a date stamp here is fine and useful
    - very unlikely to come from tarball
* would it be an idea to include a file ```release_version``` (or similar name) that is one line with just the stable version number? And do not include it in master branch. That way we could probably get away from hand-editing conf.py (albeit I have no problem doing that)
* would it make sense to create a small docker container that contains the currently "approved" sphinx version, so that users can create doc without the need to modify their system config?

As you see, nothing thought out, but maybe still helpful...

**Comment by deoren** [comment]
> @rgerhards: @deoren Unfortunately I will not be able to do a fresh test today, but let me at least give my personal opinion (maintainer hat off ;-))

Alright.

> @rgerhards: * The web server script I used for the rsyslog site is very simplistic --  a git checkout followed by sphinx-build. Maybe a short doc on what to use instead might be useful not only for me.

Is that for the stable release or master?

> @rgerhards:  * I think we should differentiate between two cases
>   - stable builds 
>       - need a plain version number, eg. 8.32.0
>       - always come from v?-stable git branch or tarball

The stable builds are covered now by way of:

- the `release_build.sh` script, which creates a tarball with filled out '8.32' values (Sphinx "version" variable) and '8.32.0' (Sphinx "release" variable). That number is calculated off of the latest git tag available at the time you run the script in your copy of the repo.
- the `source/conf.py` included in release tarballs (going forward that is). You just run `sphinx-build -b html source build` as the README directions indicate.
- running `sphinx-build -D release="8.32.0" -D version="8.32" -b html source build` manually if you want to override the values. This is the approach you would use if you wish to generate from the v8-stable branch and wish to specify a specific version. My thoughts were that you would use your `release_build.sh` script and then rsync over the built content to the main website. You can just as easily run the same script directly on the web server however, or with some minor changes pass in as arguments to the script where you would like the content to go. I would probably write a separate script that "drives" the `release_build.sh` script and then places the output (via rsync or another method) where you wish it to go (the tarball and the `build` directory content).

> @rgerhards:  
>   - dev (master) builds
>     - IMHO a date stamp here is fine and useful
>     - very unlikely to come from tarball 

I completely agree on all points, but I'm definitely biased in that dev builds help me when reviewing the docs before/after making changes (especially epub format for longer reviews). Anything that makes that process easier is something I'm likely to argue for. :) On a tangent, but building all branches via my VPS is specially helpful for this work.

For those builds, the current state of this branch automatically calculates all values:

- http://rsyslog.whyaskwhy.org/display_branch_build_info/
    - top-left 
        - *rsyslog 8.33-display_branch_build_info-20180115-7c1abe1 documentation*
    - bottom
        - *Built on Jan 15, 2018 from branch display_branch_build_info, commit 7c1abe1*

This occurs whenever someone simply runs `sphinx-build -b html source build` from a clone of the repo. Our README file covers this case. If the users run the same command from the tarball, they will get a copy of the stable docs that match that version (without dev build "decoration").

> @rgerhards:  
> * would it be an idea to include a file ```release_version``` (or similar name) that is one line with just the stable version number? And do not include it in master branch. That way we could probably get away from hand-editing conf.py (albeit I have no problem doing that)

We could, but the changes in this branch negate a need to specify the version number for dev or stable builds. I think you'll see what I mean when you have a chance to run through the steps I outlined (which admittedly does take a little time in order to experience for yourself what I'm referring to).

> @rgerhards:  
> * would it make sense to create a small docker container that contains the currently "approved" sphinx version, so that users can create doc without the need to modify their system config?

Perhaps, but I don't have docker experience (yet), so I can't say for sure. I would think that running `pip install -r requirements.txt` would be easier than setting up docker in order to run a test build. I don't know docker yet, so I would personally opt for using pip since there are already so many sites dedicated to walking users through the task (which our README now also covers to some extent).

Our README also walks them through setting up a virtual environment to build the docs, so that isolates some of the changes (other than installing `virtualenv` package, which other guides will also ask the user to do). I think the docker image has value and is worth doing, but as an enhancement. Users can download it, boot it up and build the docs. All in one go. But, they could just as easily grab the docs from rsyslog.com (or readthedocs.org, once I can make some time to follow-up and contact the owner of that repo again).

Another tangent topic is to setup an automated build system that regularly builds all branches in all major formats for easy access. My VPS and the work I'm doing there is a prototype for what I'd like to see this project do in the future. I assume the interest is low right now for frequent builds, so I will use my VPS to demo content changes. I'll also offer doing the same for others at some point once I have time to work through the security aspect.

> @rgerhards:  As you see, nothing thought out, but maybe still helpful...

I think it's useful, provided that you don't use it as an excuse to skip the chore list I've assigned you. ;)

**Comment by davidelang** [comment]
On Mon, 15 Jan 2018, Rainer Gerhards wrote:

> * I think we should differentiate between two cases
>  - stable builds
>      - need a plain version number, eg. 8.32.0
>      - always come from v?-stable git branch or tarball
>  - dev (master) builds
>    - IMHO a date stamp here is fine and useful
>    - very unlikely to come from tarball

using git-describe may be more useful for dev builds than a date (when was the 
git tree last synced with master? the date doesn't tell you)

how about something the detects if git is there, and if not prompts for a 
version?

**Comment by deoren** [comment]
>@rgerhards:
>
> * I think we should differentiate between two cases
>  - stable builds
>      - need a plain version number, eg. 8.32.0
>      - always come from v?-stable git branch or tarball
>  - dev (master) builds
>    - IMHO a date stamp here is fine and useful
>    - very unlikely to come from tarball

> @davidelang: using git-describe may be more useful for dev builds than a date (when was the 
> git tree last synced with master? the date doesn't tell you)

There is a function in this branch (and a subprocess call in other branches aside from current master) that use it to retrieve the current commit hash for reference, but not for any of the other details provided by that command. I looked up `git describe` and found this explanation:

> The command finds the most recent tag that is reachable from a commit. If the tag points to the commit, then only the tag is shown. Otherwise, it suffixes the tag name with the number of additional commits on top of the tagged object and the abbreviated object name of the most recent commit.

That is useful, but when working with a dev build of the docs knowing how far off from a recent tag (presumably the recent stable release?) doesn't help me a lot. As it is, running the command for this branch doesn't tell me much:

`$ git describe`
```
v8.29.0-227-g7c1abe1
```

I can tell we are 227 commits from the v8.29.0 annotated tag, but it is hard to get a sense of when it was built. If we opt to use any tag for a reference point, we find this:

`$ git describe --tags`
```
v8.32.0-24-g7c1abe1
```

24 commits since the last stable release.

I suppose pulling the date from the last commit is more useful than the "last generated" date.

`$ git log -n 1 --date=short | grep Date | awk '{print $NF}'`
```
2018-01-15
```

That could be useful; swap out "date generated" for "date branch last updated".

> how about something the detects if git is there, and if not prompts for a 
> version?

My last response (https://github.com/rsyslog/rsyslog-doc/pull/493#issuecomment-357795879) goes into more detail, but in short:

- the current `release_build.sh` script
    - bails out if not run from a Git repo
    - otherwise if run from a Git repo, it modifies the `source/conf.py` file to include static `version = '8.32'` and `release = '8.32.0'` lines and this is included in the tarball

- running `sphinx-build -b html source build`
    - if from the release tarball, this generates a stable release for the tag it is associated with (8.32.0 tarball results in 8.32.0 docs)
    - if run from the repo, the `source/conf.py` Sphinx build conf will use the local Git repo to determine the build info for the current branch
    - if from the tarball downloaded from GitHub, I don't know. I suspect it may have the problem previously reported on #504. I'll test and see. Falling back to a prompt could work, but we could just as well fall back to a static string instead. Something like "rsyslog dev-build" or something similar. I suspect that most users wishing to build the docs will do so either from the repo or from the release tarball, but having a good fallback makes sense.

**Comment by deoren** [comment]
> @deoren:
> - running `sphinx-build -b html source build`
>     - if from the release tarball, this generates a stable release for the tag it is associated with (8.32.0 tarball results in 8.32.0 docs)
>     - if run from the repo, the `source/conf.py` Sphinx build conf will use the local Git repo to determine the build info for the current branch
>     - if from the tarball downloaded from GitHub, I don't know. I suspect it may have the problem previously reported on #504. I'll test and see. Falling back to a prompt could work, but we could just as well fall back to a static string instead. Something like "rsyslog dev-build" or something similar. I suspect that most users wishing to build the docs will do so either from the repo or from the release tarball, but having a good fallback makes sense.

When downloading the zip file, I was reminded that the `master` branch is the default. Per rsyslog/rsyslog-doc#505, the `master` branch has a hard-coded version number. That was intended as a workaround to solve the immediate problem of a failed "bare" build when using a non-Git repo as a build source until this branch could be sorted and approved.

Therefore `master` currently builds without issue from any source, be it Git repo or downloaded zip file from GitHub.

It's probably a good idea to use 'master' as the release variable placeholder value instead of the hard-coded strings that I have in now. This will allow a meaningful build value for when Git isn't available and still is something that can be matched against. For the `version` build conf variable, I'm not sure yet. I'll have to give it some thought in order to have something useful to propose. Perhaps just '8', which would result in 'rsyslog 8 master' when the docs are generated from a zip file build. In those cases it doesn't have to look pretty, just get across that it is from the `master` branch, that it is for v8 series and of course that it is documentation for rsyslog.

I'll make those changes in the next day and test.

**Comment by deoren** [comment]
86ba528 has the changes in. Thus far a zip file build works as expected, and I expect that existing builds as noted previously for this branch will be the same. I'll test when I'm able and report back.

**Comment by rgerhards** [comment]
Posting to the main thread as the review will be hidden after an amend: I would appreciate if we could also amend the README with clear instructions (like currently given for sphinx-build) to:

* generate a single-branch build directory for e.g. consumption on a web server
* what to do to generate a release tarball
* where exactly version numbers need to be manually preset. IMHO we should have some place that we update once on release and that contains the exact version number to be used by this build - no matter if building from git or tarball

In my personal opinion it would make sense to have the next version number, plus date & latest commit hash in the master doc. I find it easier to see how current it is. But I do not have a very strong need to have it handled that way.

**Comment by rgerhards** [comment]
Another side-note, just a thought. There is good information inside this thread. Maybe it would make sense to not write it in this thread, but craft it as suggested README content. But that may also be a bad idea ;-) My intent is to get to a merge with all necessary information quickly after agreement. But, once again, not sure if this is the right vehicle. Still wanted to let the idea float around...

**Comment by davidelang** [comment]
On Tue, 16 Jan 2018, Deoren Moor wrote:

>> @davidelang: using git-describe may be more useful for dev builds than a date (when was the
>> git tree last synced with master? the date doesn't tell you)
>
> There is a function in this branch (and a subprocess call in other branches aside from current master) that use it to retrieve the current commit hash for reference, but not for any of the other details provided by that command. I looked up `git describe` and found this explanation:
>
>> The command finds the most recent tag that is reachable from a commit. If the tag points to the commit, then only the tag is shown. Otherwise, it suffixes the tag name with the number of additional commits on top of the tagged object and the abbreviated object name of the most recent commit.
>
> That is useful, but when working with a dev build of the docs knowing how far off from a recent tag (presumably the recent stable release?) doesn't help me a lot. As it is, running the command for this branch doesn't tell me much:
>
> `$ git describe`
> ```
> v8.29.0-227-g7c1abe1
> ```
>
> I can tell we are 227 commits from the v8.29.0 annotated tag, but it is hard to get a sense of when it was built. If we opt to use any tag for a reference point, we find this:
>
> `$ git describe --tags`
> ```
> v8.32.0-24-g7c1abe1
> ```

Thanks, that is a trick I didn't realize was there, that helps.

> 24 commits since the last stable release.
>
> I suppose pulling the date from the last commit is more useful than the "last generated" date.
>
> `$ git log -n 1 --date=short | grep Date | awk '{print $NF}'`
> ```
> 2018-01-15
> ```
>
> That could be useful; swap out "date generated" for "date branch last updated".

number of commits since a known point is more useful than the date of that 
commit. Remember that you don't know if those commits are in the upstream or 
not. All you know is the last tagged point in common was v8.32.0, those 
additional commits may be upstream, or may be local. But you know that this is 
not the 8.32 release docs, and that there has been some changes, but not tons of 
changes since the release.

And if development gets really hectic, you could make -rc tags every week or 
two.

>> how about something the detects if git is there, and if not prompts for a
>> version?
>
> My last response (https://github.com/rsyslog/rsyslog-doc/pull/493#issuecomment-357795879) goes into more detail, but in short:
>
> - the current `release_build.sh` script
>    - bails out if not run from a Git repo
>    - otherwise if run from a Git repo, it modifies the `source/conf.py` file to include static `version = '8.32'` and `release = '8.32.0'` lines and this is included in the tarball

tagging things with git describe --tags would be  bit better. It would make no 
difference if building from a tagged release, but would identify if there are 
changes since the last tag.

**Comment by rgerhards** [comment]
> number of commits since a known point is more useful than the date of that
> commit. Remember that you don't know if those commits are in the upstream or
> not. All you know is the last tagged point in common was v8.32.0, those
> additional commits may be upstream, or may be local. But you know that this is
> not the 8.32 release docs, and that there has been some changes, but not tons of
> changes since the release.

Thinking over it again, I think the most important information is that it is master branch, aka *unreleased*. That means folks need to use it together with today's master branch (we definitely do not support each and every commit, just master HEAD!). The date would just be a checkmark for me that nothing is stalled, but IMHO does not provide really useful information to users. Nor does changes-since-tag, nor commit hash, ...

**Comment by davidelang** [comment]
for docs it may not make sense to worry about the specific commit or how far you 
are past a tag. I admit I'm thinking in more general terms and the ability to 
re-use this on the other sub-projects

**Comment by rgerhards** [comment]
> for docs it may not make sense to worry about the specific commit or how far you
> are past a tag. I admit I'm thinking in more general terms and the ability to
> re-use this on the other sub-projects

IMHO it doesn't matter. The main characteristic of "master" is that it is unstable and the current development target can change every minute. During bug reporting, though, the master HEAD commit hash can be interesting. But really nothing else.

**Comment by deoren** [comment]
> @rgerhards: Posting to the main thread as the review will be hidden after an amend: I would appreciate if we could also amend the README with clear instructions (like currently given for sphinx-build) to:
> 
> * generate a single-branch build directory for e.g. consumption on a web server
> * what to do to generate a release tarball
> * where exactly version numbers need to be manually preset. IMHO we should have some place that we update once on release and that contains the exact version number to be used by this build - no matter if building from git or tarball
> 
> In my personal opinion it would make sense to have the next version number, plus date & latest commit hash in the master doc. I find it easier to see how current it is. But I do not have a very strong need to have it handled that way.

Instead of using the README that displays on the index of the project (which I think of as intended for newcomers), what about a README intended just for regular contributors? Perhaps CONTRIB_README.md or BUILDS_README.md?

I agree that writing up some basic coverage of the process for our use would be good. I will craft a prototype doc to go along with this branch. I'll probably go with the second proposed filename for now just to get started.

>> @deoren: My last response (https://github.com/rsyslog/rsyslog-doc/pull/493#issuecomment-357795879) goes into more detail, but in short:
>>
>> - the current `release_build.sh` script
>>    - bails out if not run from a Git repo
>>    - otherwise if run from a Git repo, it modifies the `source/conf.py` file to include static `version = '8.32'` and `release = '8.32.0'` lines and this is included in the tarball
>
> @davidelang: tagging things with git describe --tags would be  bit better. It would make no difference if building from a tagged release, but would identify if there are changes since the last tag.

The `version` and `release` Sphinx build config variables that I mentioned are used during the build process for the docs and are prominently displayed. For the version of the docs built/distributed by distro maintainers, they will probably wish to have the docs display the release string as 'rsyslog 8.32.0'. This will display in the top-left corner as 'rsyslog 8.32.0 documentation'.

Setting those static values (from what is current at the time you run it) is what the `release_build.sh` script does just before including the just modified `source/conf.py` file in the tarball.

> @rgerhards: Thinking over it again, I think the most important information is that it is master branch, aka unreleased. That means folks need to use it together with today's master branch (we definitely do not support each and every commit, just master HEAD!). The date would just be a checkmark for me that nothing is stalled, but IMHO does not provide really useful information to users. Nor does changes-since-tag, nor commit hash, ...

The date was for me the same thing: how long since the docs were last generated. If I wanted to know for sure what changes are in the docs I would use the commit hash to match against log entries. This is particularly useful to me for non-HTML builds (epub, PDF) so I can look at the comments/highlights I have made and will have a general idea of whether I've created issues for the FIXME marks I've made in my review copies.

It's also useful so I can tell at a glance whether the cron script for rsyslog.whyaskwhy.org has hiccuped and fresh builds are not being generated as they should.

Since the cron script is currently building ALL branches for a repo, the combination of the details I mentioned are useful (at least to me) to help with having others on the docs review the material. They can take a quick look at a prototype and (once folks are used to the details being there) can mention the branch and commit that they have a concern about.

> > @davidelang: for docs it may not make sense to worry about the specific commit or how far you are past a tag. I admit I'm thinking in more general terms and the ability to re-use this on the other sub-projects

> @rgerhards: IMHO it doesn't matter. The main characteristic of "master" is that it is unstable and the current development target can change every minute. During bug reporting, though, the master HEAD commit hash can be interesting. But really nothing else.

The specific commit was intended to aid in several situations, but predominantly for my own use when reviewing doc builds (I use Google Docs a lot and the ability to sync highlights and annotations are of huge benefit to me). Sometimes I will review a copy over a length of time and when I go back over the remarks (so I can update or create new issues here) it is nice to quickly/easily know which build I was reviewing vs the latest master branch.

As I mentioned a little further up in this reply, the commit hash and date are intended to be useful tools for others on the doc team when we review prototype changes as a group.

**Comment by deoren** [comment]
> @deoren: 
>
> Instead of using the README that displays on the index of the project (which I think of as intended for newcomers), what about a README intended just for regular contributors? Perhaps CONTRIB_README.md or BUILDS_README.md?
> 
> I agree that writing up some basic coverage of the process for our use would be good. I will craft a prototype doc to go along with this branch. I'll probably go with the second proposed filename for now just to get started.

I have a first draft pushed to my repo. Start with the main README to refresh your memory of what is included, then click on the new link to the BUILDS_README.md file and let me know what you think.

I've not given it a solid review for grammatical issues, so there is likely some additional work needed.

Either of these URLs should work for you:

- https://github.com/deoren/rsyslog-doc/blob/display_branch_build_info/README.md
- https://github.com/deoren/rsyslog-doc/tree/display_branch_build_info/

**Comment by deoren** [comment]
> > @deoren: 
> >
> > Instead of using the README that displays on the index of the project (which I think of as intended for newcomers), what about a README intended just for regular contributors? Perhaps CONTRIB_README.md or BUILDS_README.md?
> > 
> > I agree that writing up some basic coverage of the process for our use would be good. I will craft a prototype doc to go along with this branch. I'll probably go with the second proposed filename for now just to get started.
 
> I have a first draft pushed to my repo. Start with the main README to refresh your memory of what is included, then click on the new link to the BUILDS_README.md file and let me know what you think.
> 
> I've not given it a solid review for grammatical issues, so there is likely some additional work needed.
> 
> Either of these URLs should work for you:
> 
> - https://github.com/deoren/rsyslog-doc/blob/display_branch_build_info/README.md
> - https://github.com/deoren/rsyslog-doc/tree/display_branch_build_info/

@rgerhards I made some light tweaks to it earlier. No rush, but when you get a chance let me know what you think. I'm happy to include some additional details for uploading/managing content on your end, providing you're comfortable publishing those details in the doc. Otherwise I can add some obfuscated steps as a reminder if you would find that useful (e.g., "After the build, run the X script on the primary web server to do Y, then ..., followed by Z").

**Comment by deoren** [comment]
@rgerhards I've squashed the commits for this PR. All checks that I've made suggest that this PR is in pretty good shape, but I'll do a final round of testing and post back this weekend with my results.

I think this is nearly ready to merge, pending your review of the results.

**Comment by deoren** [comment]
@mbiebl Provided I didn't overlook anything, 62838dfa0b585a0a341a66637798582bb7e77ec6 will probably be merged soon. Do you see any potential issues on your end as a result of the changes?

The short of it is:

- If the build is run without the new placeholder values then the provided values (e.g., stable build so the stable release version number) are used to generate the build
- If the build is run with the placeholder values, then a check is made to locate the .git directory
    - If it is there, the version/release values are dynamically generated from the Git repo
    - If it is NOT there, the placeholder values are used for the build

In all cases, Git is an enhancement to the build process, not a default dependency.

**Comment by deoren** [comment]
@rgerhards I'm ready to merge this in as well. I have some more work planned for the cron script, but I am already planning to do further testing and handle those changes as part of a later PR.

**Comment by deoren** [comment]
@rgerhards I've gone ahead and merged this. I've been performing regular test builds from the results of the PR, but will proceed to perform some follow-up tests now that the changes are in. I'll submit further PRs as necessary to correct any mistakes that might have crept in.

**Comment by rgerhards** [comment]
excellent - I'll also try to finally update our own automated build and will exactly follow the guidance as a double check.
